### PR TITLE
Deprecate xshift (X) and yshift (Y) aliases from all plotting modules (remove in v0.12.0)

### PR DESCRIPTION
**Pull request by michaelgrund**
**Description of proposed changes**

This PR addresses #924 and removes the xshift (X) and yshift (Y) aliases from all plotting modules and gallery examples.

Modules:

* [x] Figure.basemap
* [x] Figure.coast
* [x] Figure.colorbar
* [x] Figure.contour
* [x] Figure.grdcontour
* [x] Figure.grdimage
* [x] Figure.grdview
* [x] Figure.histogram
* [x] Figure.image
* [x] Figure.legend
* [x] Figure.logo
* [x] Figure.meca
* [x] Figure.plot
* [x] Figure.plot3d
* [x] Figure.rose
* [x] Figure.solar
* [x] Figure.subplot
* [x] Figure.text
* [x] Figure.velo
* [x] Figure.wiggle

Gallery examples:

* [x] scatter3d.py

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by michaelgrund** [comment]
Should we add the warning/error if xshift, yshift, X or Y are passed directly into each function or create a correspondig decorator @GenericMappingTools/pygmt-maintainers? Exemplary I added one in basemap.py.

**Comment by seisman** [comment]
> Should we add the warning/error if xshift, yshift, X or Y are passed directly into each function or create a correspondig decorator @GenericMappingTools/pygmt-maintainers? Exemplary I added one in basemap.py.

I feel it's possible to add the warning in the `use_alias` decorator, instead of creating a new decorator.

**Comment by michaelgrund** [comment]
> > Should we add the warning/error if xshift, yshift, X or Y are passed directly into each function or create a correspondig decorator @GenericMappingTools/pygmt-maintainers? Exemplary I added one in basemap.py.
> 
> I feel it's possible to add the warning in the `use_alias` decorator, instead of creating a new decorator.

I updated the  `use_alias` decorator. Is this the direction you had in mind @seisman?

**Comment by seisman** [comment]
> > > Should we add the warning/error if xshift, yshift, X or Y are passed directly into each function or create a correspondig decorator @GenericMappingTools/pygmt-maintainers? Exemplary I added one in basemap.py.
> > 
> > 
> > I feel it's possible to add the warning in the `use_alias` decorator, instead of creating a new decorator.
> 
> I updated the `use_alias` decorator. Is this the direction you had in mind @seisman?

Yes. Does it work as expected?

File: pygmt/helpers/decorators.py
Comment (seisman) [inline pygmt/helpers/decorators.py:568]: I think it doesn't works, because after removing `X="xshift"` and `Y="yshift"` from the `use_alias` decorator, the variable `aliases` at line 571 no longer contain the paramters `xshift` or `yshift`.

Instead, after line 583, you can check if any of `xshift`, `yshift`, `X` and `Y` exists in `kwargs`. If yes, then raise an warning. For backward compatibility, you also need to do the "xshift"->X and "yshift"->"Y" conversions if "xshift" or "yshift" is used.

Comment (michaelgrund) [reply pygmt/helpers/decorators.py:568]: So, you mean to leave `X="xshift" and Y="yshift"` as is in each function and add the others at this point, too @seisman ?
```
@use_alias(
    X="xshift",
    Y="yshift",
    xshift="X",
    yshift="Y",
)
```

Comment (seisman) [reply pygmt/helpers/decorators.py:568]: > So, you mean to leave `X="xshift" and Y="yshift"` as is in each function and add the others at this point, too @seisman ?
> 
> ```
> @use_alias(
>     X="xshift",
>     Y="yshift",
>     xshift="X",
>     yshift="Y",
> )
> ```

No. Changes to other files (e.g., `basemap.py`) make sense to me. The only thing that is missing is how to _correctly_ check if users use `xshift`, `yshift`, `X` or `Y`. The checking should be done in the `pygmt/helpers/decorators.py` file.

File: pygmt/helpers/decorators.py
Comment (weiji14) [inline pygmt/helpers/decorators.py:575]: Using an error here will immediately break people's code if they're using `xshift` or `yshift`. Better to go through a deprecation cycle that raises a `FutureWarning` instead.

One idea is to use the [`@deprecate_parameter`](https://github.com/GenericMappingTools/pygmt/blob/d80b986a3f5634e2e68d98d569405dfa3eba1475/pygmt/helpers/decorators.py#L732-L808) function and do something like `@deprecate_parameter(oldname="xshift", newname=None, deprecate_version="v0.8.0", remove_version="0.10.0")`. This will require some modifications to the decorator to handle `newname=None` though.

Comment (seisman) [reply pygmt/helpers/decorators.py:575]: > One idea is to use the [`@deprecate_parameter`](https://github.com/GenericMappingTools/pygmt/blob/d80b986a3f5634e2e68d98d569405dfa3eba1475/pygmt/helpers/decorators.py#L732-L808) function and do something like `@deprecate_parameter(oldname="xshift", newname=None, deprecate_version="v0.8.0", remove_version="0.10.0")`. This will require some modifications to the decorator to handle `newname=None` though.

We don't have to use `@deprecate_parameter` for this, which requires adding `@deprecate_parameter` decorators in many Python files.

Comment (michaelgrund) [reply pygmt/helpers/decorators.py]: I modified the passage @seisman. I understand the concerns @weiji14 has regarding raising an error here. However, we did agree to totally drop this option (not only deprecate it), didn't we?

Comment (seisman) [reply pygmt/helpers/decorators.py]: > was there a discussion to drop xshift/yshift in a backward incompatible way?

No, we definitely need **backward compatibility**.

The final goal is, for the next 2 or more minor versions, users can still use parameters `xshift`, `yshift`, `X`, `Y`, and they will see a *FutureWarning* like:
```
Parameters xshift, yshift, X and Y are deprecated and will be no longer 
supported in v0.x.x. Please user Figure.shift_origin() instead.
```

> One idea is to use the [`@deprecate_parameter`](https://github.com/GenericMappingTools/pygmt/blob/d80b986a3f5634e2e68d98d569405dfa3eba1475/pygmt/helpers/decorators.py#L732-L808) function and do something like `@deprecate_parameter(oldname="xshift", newname=None, deprecate_version="v0.8.0", remove_version="0.10.0")`. This will require some modifications to the decorator to handle `newname=None` though.

I believe this will work but it's not ideal. It's easy to modify the `deprecate_parameter` decorator to support `newname=None`, but the warning message would be something like:
```
The 'xxxx' parameter has been deprecated since vx.x.x and will be removed in vx.x.x.
```
The error message can't recommend using `Figure.shift_origin()` unless we deal with `xshift` and `yshfit` in a special way.

It's also more confusing when users use the short parameters `X` or `Y`. When they use `X`, they are recommended to use `xshift`. After they modify `X` to `xshift`, then they see `xshfit` is deprecated.

**Comment by seisman** [comment]
@michaelgrund There are a lot of conflicts after merging PR #1932 into the main branch. Please resolve these conflicts first.

File: pygmt/src/solar.py
Comment (seisman) [inline pygmt/src/solar.py:57]:
```suggestion
        is ``"0.25p,black,solid"``.
```

File: pygmt/helpers/decorators.py
Comment (seisman) [reply pygmt/helpers/decorators.py]: I've opened PR #2135, which shows how we can deprecate `timestamp` (`U`) in a backward-compatible and recommend `Figure.timestamp()` (not implemented yet) instead. The same idea can be applied to xshift and yshift in this PR.

Comment (michaelgrund) [reply pygmt/helpers/decorators.py]: I applied your implemetations for this PR @seisman.

File: pygmt/helpers/decorators.py
Comment (seisman) [inline pygmt/helpers/decorators.py:586]: Do we want to fully remove these parameters in future versions (e.g., v0.12.0)?
```suggestion
                msg = (
                    "Parameters 'X' and 'xshift' are deprecated since v0.8.0 "
                    "and will be removed in v0.12.0. "
                    "Use Figure.shift_origin(xshift=...) instead."
                )
```

**Comment by seisman** [comment]
@michaelgrund I think this PR is almost ready for the final review. What do you think?

File: pygmt/helpers/decorators.py
Comment (weiji14) [reply pygmt/helpers/decorators.py:586]: 4 minor versions sounds good.

**Comment by michaelgrund** [comment]
> @michaelgrund I think this PR is almost ready for the final review. What do you think?

Yes it is!

**Comment by seisman** [comment]
Some changes need to be made:

- [ ] Need to remove `xshift` and `yshift` from https://github.com/GenericMappingTools/pygmt/blob/main/pygmt/src/ternary.py
- [ ] `xshift` and `yshift` are still used in https://github.com/GenericMappingTools/pygmt/blob/main/pygmt/tests/test_config.py.
- [ ] Placeholder `xyshift` is no longer used and can be removed from COMMON_DOCSTRING: https://github.com/GenericMappingTools/pygmt/blob/6964fe13cd1463452457becf4221a80387b674f6/pygmt/helpers/decorators.py#L100
- [ ] Need to add a test to make sure that a warning is raised if users use xshift/X/yshift/Y

**Comment by michaelgrund** [comment]
> Some changes need to be made:
> 
> * [x]  Need to remove `xshift` and `yshift` from https://github.com/GenericMappingTools/pygmt/blob/main/pygmt/src/ternary.py
> * [x]  `xshift` and `yshift` are still used in https://github.com/GenericMappingTools/pygmt/blob/main/pygmt/tests/test_config.py.
> * [x]  Placeholder `xyshift` is no longer used and can be removed from COMMON_DOCSTRING: https://github.com/GenericMappingTools/pygmt/blob/6964fe13cd1463452457becf4221a80387b674f6/pygmt/helpers/decorators.py#L100
> * [x]  Need to add a test to make sure that a warning is raised if users use xshift/X/yshift/Y

Added 4 tests covering each parameter. Do we need all 4 or is one sufficient @seisman ?

**Comment by seisman** [comment]
> > Some changes need to be made:
> > 
> > * [x]   Need to remove `xshift` and `yshift` from [`main`/pygmt/src/ternary.py](https://github.com/GenericMappingTools/pygmt/blob/main/pygmt/src/ternary.py?rgh-link-date=2022-11-16T20%3A33%3A26Z)
> > * [x]   `xshift` and `yshift` are still used in [`main`/pygmt/tests/test_config.py](https://github.com/GenericMappingTools/pygmt/blob/main/pygmt/tests/test_config.py?rgh-link-date=2022-11-16T20%3A33%3A26Z).
> > * [x]   Placeholder `xyshift` is no longer used and can be removed from COMMON_DOCSTRING: https://github.com/GenericMappingTools/pygmt/blob/6964fe13cd1463452457becf4221a80387b674f6/pygmt/helpers/decorators.py#L100
> > * [x]   Need to add a test to make sure that a warning is raised if users use xshift/X/yshift/Y
> 
> Added 4 tests covering each parameter. Do we need all 4 or is one sufficient @seisman ?

I think we can merge the four tests into one larger test.

File: pygmt/tests/test_figure.py
Comment (willschlitzer) [inline pygmt/tests/test_figure.py:298]: Does the testing of `X` and `Y` needs to be included if `xshift` and `yshift` are already included, since we trust that the aliases translate to the appropriate arguments?

Comment (seisman) [reply pygmt/tests/test_figure.py:298]: > Does the testing of `X` and `Y` needs to be included if `xshift` and `yshift` are already included, since we trust that the aliases translate to the appropriate arguments?

I think we still need to test `X` and `Y`, because we have removed the "X=>xshift" and "Y=>yshift" aliases in this PR.

File: pygmt/tests/test_figure.py
Comment (seisman) [inline pygmt/tests/test_figure.py:320]:
```suggestion
def test_figure_deprecated_xshift_yshift():
    """
    Check if deprecation of parameters X/Y/xshift/yshift work correctly if used.
    """
    fig = Figure()
    fig.basemap(region=[0, 1, 0, 1], projection="X1c/1c", frame=True)
    with pytest.warns(expected_warning=SyntaxWarning) as record:
        fig.plot(x=1, y=1, style="c3c", xshift="3c")
        assert len(record) == 1  # check that only one warning was raised
    with pytest.warns(expected_warning=SyntaxWarning) as record:
        fig.plot(x=1, y=1, style="c3c", X="3c")
        assert len(record) == 1  # check that only one warning was raised
    with pytest.warns(expected_warning=SyntaxWarning) as record:
        fig.plot(x=1, y=1, style="c3c", yshift="3c")
        assert len(record) == 1  # check that only one warning was raised
    with pytest.warns(expected_warning=SyntaxWarning) as record:
        fig.plot(x=1, y=1, style="c3c", Y="3c")
        assert len(record) == 1  # check that only one warning was raised
```

**Review by seisman — APPROVED** [review]

**Comment by seisman** [comment]
Ping @GenericMappingTools/pygmt-maintainers for final reviews.

**Review by willschlitzer — APPROVED** [review]
Nice work @michaelgrund!

**Pull request merged**
